src/gevent/pool.py

```diff
@@ -714,19 +717,39 @@ def free_count(self):
             return 1
         return max(0, self.size - len(self))
 
-    def add(self, greenlet):
+    def add(self, greenlet, blocking=True, timeout=None):
         """
-        Begin tracking the given greenlet, blocking until space is available.
+        Begin tracking the given greenlet, possibly blocking until space is
+        available.
+
+        :keyword bool blocking: If True (the default), this function will block
+        until the pool has space or a timeout occurs.
+        :keyword float timeout: The maximum number of seconds this method will block.
+
+        :return: True if the greenlet was added; False if a Timeout occured or
+        if blocking is True and the pool is full.
 
         .. seealso:: :meth:`Group.add`
         """
-        self._semaphore.acquire()
+        try:
+            # NOTE: The docs for Semaphore.acquire indicate that it may raise a Timeout rather
```

Comment:
A `Timeout` error can be raised if a higher caller had set a timer and it expired. The Semaphore itself won't raise a Timeout. So that must be ignored here and just allowed to propagate (that is, the try/except wrapping can go away).

The pattern for working with timeouts is:
```python
timeout = Timeout(...)
try:
   do_some_stuff()
except Timeout as e:
   if e is not timeout: 
     # Somebody that called us had a timer set. It expired, not ours. 
     # They need to know about it, so this must raise
      raise
  # Our timer expired. Do whatever we do in that case.
```

Reply:
Ah, thanks for the explanation of Semaphore's timeout logic.

I agree that it's cleaner to have `Pool.add` raise a `Timeout` rather than return True/False. Will modify.